src/libs/API/commands/BaseAPICommandBlocking.js

```diff
@@ -0,0 +1,44 @@
+import BaseAPICommand from './BaseAPICommand';
+import Onyx from 'react-native-onyx';
+import ONYXKEYS from '../../../ONYXKEYS';
+import CONST from '../../../CONST';
+
+/**
+ * API Commands that need to have some kind of blocking UI while the network request is being made
+ * can extend this class.
+ */
+export default class extends BaseAPICommand {
+    // Use withOnyx to access the data about this blocking request (eg. ValidateEmailCommand.blockingCommandOnyxKey)
+    static blockingCommandOnyxKey = `BlockingCommand-${this.commandName}`;
+
+    /**
+     * An abstract method to define what should happen before the network request is made
+     */
+    startBlocking() {
+        Onyx.merge(this.blockingCommandOnyxKey, {loading: true, error: null});
+    }
+
+    /**
+     * An abstract method to define what should happen after the network request is made
+     */
+    finishBlocking() {
+        Onyx.merge(this.blockingCommandOnyxKey, {loading: false});
+    }
+
+    /**
+     * Override the base makeRequest method to add the blocking functionality
+     *
+     * @param {Object} parameters
+     * @return {Promise<unknown>}
+     */
+    makeRequest(parameters) {
+        this.startBlocking();
+        return super.makeRequest(parameters).finally(this.finishBlocking);
+    }
+
+    requestFailed(response) {
+        super.requestFailed(response);
```

Comment:
Thinking out loud, I think we should either go "down the hierarchy" or "up the hierarchy" but not both, because I think that's not necessary and it makes the code harder to follow.

I haven't put too much thought on this particular usage, but I think "Down the hierarchy" would be better here.

Down the hierarchy means calling methods of the subclasses in the parent class, like `makeRequest` in `BaseAPICommand` is calling `requestFailed`. In this design we'd put common logic in the parent class, and specific logic in the subclasses, and then in the parent class we call methods that will have specific functionality for the different cases represented by the subclasses.

Up the hierarchy means calling parent class methods from subclasses, like we're doing here in `requestFailed` calling `super.requestFailed`. 

- BaseAPICommandBlocking - makeRequest
- BaseAPICommandBlocking - startBlocking
- BaseAPICommandBlocking - super.makeRequest
- BaseAPICommand - makeRequest (let's assume it fails)
- BaseAPICommand - requestFailed - because it's defined there and it doesn't call the subclass method, the subclass method won't be called

One idea to change this would be to define startBlocking/finishBlocking in the base api command class as no-ops, and define them in the base blocking api command class with the actual logic

Reply:
Is the idea I proposed here similar to what you are proposing here? https://github.com/Expensify/App/pull/8316#discussion_r835534006